### PR TITLE
Add static session tests for object

### PR DESCRIPTION
**Pull request by abereziny**
Signed-off-by: Andrey Berezin <a.berezin@yadro.com>

Issues: 
nspcc-dev/neofs-node/issues/1878
nspcc-dev/neofs-node/issues/2028
nspcc-dev/neofs-node/issues/2029
nspcc-dev/neofs-node/issues/2030

### CODE REVIEW

**Review by anikeev-yadro — APPROVED** [review]

File: pytest_tests/testsuites/session_token/test_static_object_session_token.py [hunk 1]
```diff
@@ -0,0 +1,624 @@
...
+            "0:10",
+            f"{object_size-10}:10",
+            f"{max_object_size - RANGE_OFFSET_FOR_COMPLEX_OBJECT}:{RANGE_OFFSET_FOR_COMPLEX_OBJECT * 2}",
+        ]
+    else:
+        return ["0:10", f"{object_size-10}:10"]
+
+
+@pytest.fixture(scope="session")
+def wallet_owner(wallet_factory: WalletFactory) -> WalletInfo:
```
[the remark this answers is not in the document]
Comment (abereziny) [reply pytest_tests/testsuites/session_token/test_static_object_session_token.py:145]: In this case WalletFactory should take care of wallet lifetime (i.e. not create new wallet for each test if it's already created)
And IMO it's not a factory responsibility, that's why I use fixtures to ensure signle instance across tests in module

File: robot/resources/lib/python_keywords/object.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                    storage_object.oid,
+                    shell,
+                    endpoint=endpoint,
+                )
+
+    while start + max_wait_time >= time():
+        try:
+            wait_for_objects()
+            return
+        except Exception as ex:
```
[the remark this answers is not in the document]
Comment (abereziny) [reply robot/resources/lib/python_keywords/object.py:49]: Used at the end of the method

File: robot/resources/lib/python_keywords/object.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    attributes: list[dict[str, str]] = None
+    oid: str = None
+    tombstone: str = None
+
+
+@allure.step("Waiting until object will be available on all nodes")
+def wait_until_objects_available_on_all_nodes(
+    storage_objects: list[StorageObjectInfo], shell: Shell, max_wait_time: int = 60
+) -> None:
+    start = time()
```
[the remark this answers is not in the document]
Comment (abereziny) [reply robot/resources/lib/python_keywords/object.py:32]: not sure, but https://docs.python.org/3/library/time.html#time.time it looks like time.time() is not affected by timezone

File: robot/resources/lib/python_keywords/wallet.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+from dataclasses import dataclass
+
+from common import ASSETS_DIR, FREE_STORAGE, WALLET_PASS
+from neofs_testlib.shell import Shell
+from neofs_testlib.utils.wallet import get_last_address_from_wallet, init_wallet
+from python_keywords.payment_neogo import deposit_gas, transfer_gas
+
+
+@dataclass
+class WalletInfo:
```
[the remark this answers is not in the document]
Comment (abereziny) [reply robot/resources/lib/python_keywords/wallet.py]: done

File: robot/resources/lib/python_keywords/wallet.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    path: str
+    containers: list[str] = None
+
+    def get_address(self) -> str:
+        """
+        Extracts the last address from wallet.
+
+        Returns:
+            The address for the wallet.
```
[the remark this answers is not in the document]
Comment (abereziny) [reply robot/resources/lib/python_keywords/wallet.py]: done

File: robot/resources/lib/python_keywords/wallet.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        self.shell = shell
+
+    def create_wallet(self) -> WalletInfo:
+        """
+        Creates new default wallet
+
+        Returns:
+            Wallet object of new wallet
+        """
+        wallet_path = os.path.join(os.getcwd(), ASSETS_DIR, f"{str(uuid.uuid4())}.json")
```
[the remark this answers is not in the document]
Comment (abereziny) [reply robot/resources/lib/python_keywords/wallet.py]: done

File: robot/resources/lib/python_keywords/wallet.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
+import os
```
[the remark this answers is not in the document]
Comment (abereziny) [reply robot/resources/lib/python_keywords/wallet.py]: done

File: robot/resources/lib/python_keywords/session_token.py [hunk 8]
```diff
@@ -50,15 +75,68 @@ def generate_session_token(owner: str, session_wallet: str, cid: str = "") -> st
...
+    logger.info(f"Got this Session Token: {session_token}")
+    with open(file_path, "w", encoding="utf-8") as session_token_file:
+        json.dump(session_token, session_token_file, ensure_ascii=False, indent=4)
+
+    return file_path
+
+
+@allure.step("Generate Session Token For Object")
+def generate_object_session_token(
+    owner: str, session_wallet: str, oids: list[str], cid: str, verb: str, lifetime: Lifetime = None
```
[the remark this answers is not in the document]
Comment (abereziny) [reply robot/resources/lib/python_keywords/session_token.py]: done

File: robot/resources/lib/python_keywords/session_token.py [hunk 9]
```diff
@@ -50,15 +75,68 @@ def generate_session_token(owner: str, session_wallet: str, cid: str = "") -> st
...
+        cid(str): container ID of the container
+        oids(list[str]): list of objectIDs to put into session
+        verb(str): verb to grant access to;
+                   Valid verbs are: GET, RANGE, RANGEHASH, HEAD, SEARCH.
+        lifetime(optional, Expiration): lifetime options for session
+    Returns:
+        (str): the path to the generated session token file
+    """
+
+    file_path = os.path.join(os.getcwd(), ASSETS_DIR, str(uuid.uuid4()))
```
[the remark this answers is not in the document]
Comment (abereziny) [reply robot/resources/lib/python_keywords/session_token.py]: done

File: robot/resources/lib/python_keywords/session_token.py [hunk 10]
```diff
@@ -50,15 +75,68 @@ def generate_session_token(owner: str, session_wallet: str, cid: str = "") -> st
...
+
+    session_wlt_content = ""
+    with open(session_wallet) as fout:
```
[the remark this answers is not in the document]
Comment (abereziny) [reply robot/resources/lib/python_keywords/session_token.py]: removed as not needed

File: robot/resources/lib/python_keywords/session_token.py [hunk 11]
```diff
@@ -50,15 +75,68 @@ def generate_session_token(owner: str, session_wallet: str, cid: str = "") -> st
...
+        verb(str): verb to grant access to;
+                   Valid verbs are: GET, RANGE, RANGEHASH, HEAD, SEARCH.
+        lifetime(optional, Expiration): lifetime options for session
+    Returns:
+        (str): the path to the generated session token file
+    """
+
+    file_path = os.path.join(os.getcwd(), ASSETS_DIR, str(uuid.uuid4()))
+
+    session_wlt_content = ""
```
[the remark this answers is not in the document]
Comment (abereziny) [reply robot/resources/lib/python_keywords/session_token.py]: removed as not needed

File: robot/resources/lib/python_keywords/session_token.py [hunk 12]
```diff
@@ -50,15 +75,68 @@ def generate_session_token(owner: str, session_wallet: str, cid: str = "") -> st
...
+    with open(session_wallet) as fout:
+        session_wlt_content = json.load(fout)
+    session_wlt = wallet.Wallet.from_json(session_wlt_content, password="")
+    pub_key_64 = base64.b64encode(bytes.fromhex(str(session_wlt.accounts[0].public_key))).decode(
```
[the remark this answers is not in the document]
Comment (abereziny) [reply robot/resources/lib/python_keywords/session_token.py]: done

File: robot/resources/lib/python_keywords/object.py [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+    raise ex
+
+
+def delete_objects(storage_objects: list[StorageObjectInfo], shell: Shell) -> None:
+    """
+    Deletes given storage objects.
+
+    Args:
+        storage_objects (list[StorageObjectInfo]): list of objects to delete
```
[the remark this answers is not in the document]
Comment (abereziny) [reply robot/resources/lib/python_keywords/object.py]: done

File: robot/resources/lib/python_keywords/object.py [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+        shell: executor for cli command
+    Returns:
```
[the remark this answers is not in the document]
Comment (abereziny) [reply robot/resources/lib/python_keywords/object.py]: done

File: robot/resources/lib/python_keywords/object.py [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+from grpc_responses import OBJECT_ALREADY_REMOVED
+from neofs_testlib.shell import Shell
+from python_keywords.neofs_verbs import delete_object, get_object, head_object
+from tombstone import verify_head_tombstone
+
+CLEANUP_TIMEOUT = 10
+
+
+@dataclass
+class StorageObjectInfo:
```
[the remark this answers is not in the document]
Comment (abereziny) [reply robot/resources/lib/python_keywords/object.py]: done

File: robot/resources/lib/python_keywords/object.py [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+    wallet: str = None
+    file_path: str = None
+    file_hash: str = None
+    attributes: list[dict[str, str]] = None
+    oid: str = None
+    tombstone: str = None
+
+
+@allure.step("Waiting until object will be available on all nodes")
+def wait_until_objects_available_on_all_nodes(
```
[the remark this answers is not in the document]
Comment (abereziny) [reply robot/resources/lib/python_keywords/object.py]: moved

File: pytest_tests/testsuites/session_token/test_static_object_session_token.py [hunk 17]
```diff
@@ -0,0 +1,624 @@
...
+@pytest.fixture(scope="session")
+def wallet_user(wallet_factory: WalletFactory) -> WalletInfo:
+    """
+    Returns wallet which will use objects from owner via static session
+    """
+    return wallet_factory.create_wallet()
+
+
+@pytest.fixture(scope="session")
+def wallet_outsider(wallet_factory: WalletFactory) -> WalletInfo:
```
[the remark this answers is not in the document]
Comment (abereziny) [reply pytest_tests/testsuites/session_token/test_static_object_session_token.py]: renamed

File: pytest_tests/testsuites/session_token/test_static_object_session_token.py [hunk 18]
```diff
@@ -0,0 +1,624 @@
...
+
+logger = logging.getLogger("NeoLogger")
+
+RANGE_OFFSET_FOR_COMPLEX_OBJECT = 200
+
+
+@allure.step("Get fresh epoch")
+def get_fresh_epoch(shell: Shell) -> int:
+    # ensure new fresh epoch to avoid epoch switch during test session
+    epoch_before = get_epoch(shell=shell)
```
[the remark this answers is not in the document]
Comment (abereziny) [reply pytest_tests/testsuites/session_token/test_static_object_session_token.py]: done

File: pytest_tests/testsuites/session_token/test_static_object_session_token.py [hunk 19]
```diff
@@ -0,0 +1,624 @@
...
+)
+from wallet import WalletFactory, WalletInfo
+
+logger = logging.getLogger("NeoLogger")
+
+RANGE_OFFSET_FOR_COMPLEX_OBJECT = 200
+
+
+@allure.step("Get fresh epoch")
+def get_fresh_epoch(shell: Shell) -> int:
```
[the remark this answers is not in the document]
Comment (abereziny) [reply pytest_tests/testsuites/session_token/test_static_object_session_token.py]: done

File: pytest_tests/testsuites/session_token/test_static_object_session_token.py [hunk 20]
```diff
@@ -0,0 +1,624 @@
...
+
+
+@pytest.fixture(
+    params=[SIMPLE_OBJ_SIZE, COMPLEX_OBJ_SIZE],
+    ids=["simple object", "complex object"],
+    # Scope session to upload/delete each files set only once
+    scope="session",
+)
+def storage_objects(
+    wallet_owner: WalletInfo, client_shell: Shell, request: FixtureRequest
```
[the remark this answers is not in the document]
Comment (abereziny) [reply pytest_tests/testsuites/session_token/test_static_object_session_token.py]: I pass only those object which function should work with. That's the idea to not give extra items.
Wallet dependency for this fixture is only owner_wallet.

File: pytest_tests/testsuites/session_token/test_static_object_session_token.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (abereziny) [reply pytest_tests/testsuites/session_token/test_static_object_session_token.py]: #415

**Comment by abereziny** [comment]
#416

File: pytest_tests/testsuites/conftest.py [hunk 21]
```diff
@@ -74,11 +80,14 @@ def check_binary_versions(request, hosting: Hosting, client_shell: Shell):
...
-    full_path = os.path.join(os.getcwd(), ASSETS_DIR)
-    shutil.rmtree(full_path, ignore_errors=True)
-    os.mkdir(full_path)
-    yield full_path
-    shutil.rmtree(full_path)
+    with allure.step("Prepare tmp directory"):
+        full_path = os.path.join(os.getcwd(), ASSETS_DIR)
+        shutil.rmtree(full_path, ignore_errors=True)
+        os.mkdir(full_path)
+        yield full_path
```
[the remark this answers is not in the document]
Comment (abereziny) [reply pytest_tests/testsuites/conftest.py:87]: This is how pytests treats fixtures. before yield is preparation, after yield is a teardown.
https://docs.pytest.org/en/6.2.x/fixture.html#yield-fixtures-recommended

File: pytest_tests/testsuites/session_token/test_static_object_session_token.py [hunk 22]
```diff
@@ -0,0 +1,636 @@
...
+)
+from steps.storage_object import delete_objects, wait_until_objects_available_on_all_nodes
+
+logger = logging.getLogger("NeoLogger")
+
+RANGE_OFFSET_FOR_COMPLEX_OBJECT = 200
+
+
+@allure.step("Ensure fresh epoch")
+def ensure_fresh_epoch(shell: Shell) -> int:
```
[the remark this answers is not in the document]
Comment (abereziny) [reply pytest_tests/testsuites/session_token/test_static_object_session_token.py:52]: #419

File: pytest_tests/testsuites/conftest.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (abereziny) [reply pytest_tests/testsuites/conftest.py]: Moved yield out of step

**Pull request merged**